### PR TITLE
store: add streamed postings reading

### PR DESCRIPTION
**Pull request by GiedriusS**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

`readIndexRange` dominates the profiles here so let's stream reading postings into `index.Postings` instead of allocating everything at once.

Work in progress.

## Verification

Existing + ad-hoc tests.

### CODE REVIEW

**Comment by yeya24** [comment]
Hi @GiedriusS, we are interested in this feature as we saw same high heap usage issue recently of this part of code.
Do we still plan to work on this or we want to approach this problem in a different way? We are happy to collaborate if you are willing to

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -2416,142 +2415,42 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
 
 		// Fetch from object storage concurrently and update stats and posting list.
 		g.Go(func() error {
-			begin := time.Now()
+			for _, p := range ptrs[i:j] {
+				ir, err := r.block.bkt.GetRange(ctx, r.block.indexFilename(), p.ptr.Start, p.ptr.End-p.ptr.Start)
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/store/bucket.go:2419]: Are we able to still send 1 request per part and create posting reader from the get range reader?
If this is not doable, I feel it is better to maybe just download postings to disk

[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/store/bucket.go:2419]: Maybe more get range requests won't impact performance, hopefully we can have some datapoints to understand the impact.

[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/store/bucket.go:2419]: Maybe we could add this under a feature flag? if it is enabled then we would send multiple requests which would mean bigger costs if using some SaaS that charges per-request but we would get constant RAM usage

**Comment by GiedriusS** [comment]
@yeya24 https://github.com/thanos-io/thanos/pull/6442 here's a similar optimization that is based on this code and our discussions on Slack that is actually feasible. #6442 is unfinished but WDYT about such optimization?

**Comment by yeya24** [comment]
@GiedriusS IIUC https://github.com/thanos-io/thanos/pull/6442 will supercede this pr? Or will we do both?

**Comment by fpetkovski** [comment]
What about the roaring bitmap approach? I feel like that one is superior since we can aggregate postings in a streaming manner. I don't know if making one or more requests per posting is sustainable at scale.

**Comment by yeya24** [comment]
> also had an idea to calculate the intersection by merging one (or a controlled number) of postings at a time, instead of maxing out the fanout.

Yeah I am thinking about the same. We know the postings length so we can just sort and start from the two smallest size of postings. Need some benchmarks though.

**Pull request closed without merging**